ctera/src/main/java/ch/cyberduck/core/ctera/CteraAclPermissionFeature.java

```diff
@@ -0,0 +1,208 @@
+            action = action.or(Permission.Action.write);
+        }
+        if(acl.get(new Acl.CanonicalUser()).contains(EXECUTEPERMISSION) || acl.get(new Acl.CanonicalUser()).contains(TRAVERSEPERMISSION)) {
+            action = action.or(Permission.Action.execute);
+        }
+        // TODO CTERA-136 only user - what about group/others?
+        return new Permission(action, Permission.Action.none, Permission.Action.none);
+    }
+
+    private final DAVSession session;
+
+    public CteraAclPermissionFeature(final DAVSession session) {
+        this.session = session;
+    }
+
+    @Override
+    public Acl getPermission(final Path file) throws BackgroundException {
+        final PathAttributes attributes = new CteraAttributesFinderFeature(session).find(file, new DisabledListProgressListener());
+        return attributes.getAcl();
+    }
+
+    @Override
+    public void setPermission(final Path file, final Acl acl) throws BackgroundException {
+        try {
+            List<Element> setProps = this.toCustomProperties(acl);
+            session.getClient().patch(new DAVPathEncoder().encode(file), setProps, Collections.emptyList(), Collections.emptyMap());
+        }
+        catch(SardineException e) {
+            throw new DAVExceptionMappingService().map("Failure to write attributes of {0}", e, file);
+        }
+        catch(IOException e) {
+            throw new DefaultIOExceptionMappingService().map(e, file);
+        }
+    }
+
+    @Override
+    public List<Acl.User> getAvailableAclUsers() {
+        return Collections.singletonList(new Acl.CanonicalUser());
+    }
+
+    @Override
+    public List<Acl.Role> getAvailableAclRoles(final List<Path> files) {
+        return allCteraCustomACLRoles;
+    }
+
+    protected List<Element> toCustomProperties(final Acl acl) {
+        final List<Element> props = new ArrayList<>();
+        if(acl == null || acl.get(new Acl.CanonicalUser()) == null) {
+            return props;
+        }
+        for(Acl.Role role : allCteraCustomACLRoles) {
+            final QName prop = toQn(role);
+            final Element element = SardineUtil.createElement(prop);
+            element.setTextContent(Boolean.toString(acl.get(new Acl.CanonicalUser()).contains(role)));
+            props.add(element);
+        }
+        return props;
+    }
+
+    @Override
+    public Acl getDefault(final EnumSet<Path.Type> type) {
+        return Acl.EMPTY;
+    }
+
+    @Override
+    public Acl getDefault(final Path file, final Local local) throws BackgroundException {
+        return Acl.EMPTY;
+    }
+
+    protected static void checkCteraRole(final Path file, final Acl.Role role) throws BackgroundException {
+        final Acl acl = file.attributes().getAcl();
+        if(acl.equals(Acl.EMPTY)) {
+            return;
+        }
+        if(!acl.get(new Acl.CanonicalUser()).contains(role)) {
+            throw new AccessDeniedException(
```

Comment:
Log with warning level about missing role.

Reply:
This method is used in `preflight`  to throw exception when the roles required for an operation are missing - do you mean log warn in addition to throwing the exception?

Reply:
fa7729176c008116db70d13ed557724031ea7dcf